SSL security enhancement

kamermans opened the pull request:
This PR removes the DES-based TLS ciphers from the Mozilla intermediate TLS configuration, which mitigates a number of threats and increases the SSL score on [Qualys SSL Test](https://www.ssllabs.com/ssltest/) to _A+_ and [CryptCheck](https://tls.imirhil.fr/) from an _F_ to a _B_.  Some SSL/TLS scanners, like [Symantec CryptoReport](https://cryptoreport.websecurity.symantec.com/checker/) downgrade the score because `nginx-proxy` allows TLSv1.0, but in my research I've found that the following browsers would not work if we remove TLSv1.0:
- Chrome < 22
- Firefox < 27
- IE < 11
- Safari < 7
- iOS < 5
- Android Browser < 5

These represent a non-insignificant market share as of September 2016 (according to my company's [industry-wide device usage trends report](https://www.scientiamobile.com/page/movr-mobile-overview-report)) so I recommend we leave TLSv1.0 in place at least until the end of 2017, or perhaps as late as the official end of life for PCI compliance, June 30, 2018.

Also, this PR adds a default 2048-bit Diffie-Hellman groups file, `/etc/nginx/dhparam.pem`, which is generated at build time.  This allows Perfect Forward Secrecy via DH, which is an advantageous security improvement for those clients that support it.  No browsers will be turned away because of this setting and individual vhosts can still use their own DH file as before.

[edit 2017-04-04]
The following systems do not support HTTPS connections to servers with a DH param over 1024-bits and would be unable to connect with the default 2048-DHE proposed by this PR:
- Java < 6
- Java 6 < update 105
- Java 7 < update 91
NOTE: neither Java 6u105 nor Java 7u91 was a public release, so only Java users with a paid subscription to Java SE Advanced or Java SE Support would have access to this update!

[end edit 2017-04-04]

I see some talk about OCSP Stapling as well, and as tempting as it is to add it here, I need to understand the ramifications of the additional resolver and external certificate fetching before I can endorse it.

This PR is currently available from docker hub as `kamermans/nginx-proxy:ssl_upgrade` for testing, built off today's `master` branch.

As far as I see, this PR solves the following issues and PRs:
- #586 
- #535 
- #562 
- PR #588
- PR #575

cridenour commented:
@kamermans Should the dhparams be built in the Dockerfile? That would mean anyone pulling this from the docker registry would get the same params - while not the end of the world, certainly isn't the ideal setup. 

Maybe something like this in docker-entrypoint.sh?

```bash
if [ ! -f /etc/nginx/dhparam.pem ]; then
    openssl dhparam -out /etc/nginx/dhparam.pem 2048
  fi
fi
```

jwilder commented:
@kamermans What do you think about @cridenour suggestion to generate a default DH key at startup?  I can't think of any downsides to it.

kamermans commented:
If I recall correctly, I did it in the Dockerfile because it took a very long time (tens of seconds or minutes) to generate.  It's not ideal to have it in the Dockerfile, but I'm not sure it's any less secure than not having it at all.  I'm not a security expert, but I would guess that knowing the dhparams in advance would only allow a malicious user to break perfect forward secrecy, but they still need to break through TLS after that.  If someone knows more than me about it, I'm all ears!

Off the top of my head, alternate ways of securing dhparams would be:
 1. Suggest the user rebuild the image.
 2. Suggest the user mount their own dhparams at `/etc/nginx/dhparam.pem`, and emit warnings on startup.
 3. Generate a less secure, but faster dhparams on startup using less than 2048 bits (not sure if this will pass the SSL/TLS checks).
 4. Generate the dhparams in the background and once it's finished, reload nginx (might be an interesting option).

dehy commented:
As the dhparams need to be generated once, I think the build time at startup is acceptable, if in a volume for surviving future redeployment of service. But I like the #4 as well!

kamermans commented:
In my environment, I could not handle a potentially 1-minute startup as I almost always `--force-recreate` my images to get rid of transient data.  Let me think about option 4 tonight...

dehy commented:
So in this thread : https://github.com/jchaney/owncloud/issues/42#issuecomment-182110124

> I choose option 2. This is what most people seem to be doing, and what most ssl configuring guides and stackexchange answers point to. And because it takes away the fear of an attack like logjam, where a part of the impact came from the fact that lots of sites where using the same primes.

I agree with him. But he also note that generating the primes in the build could also be a viable option. So that's your call :D

kamermans commented:
Ok, so here are my findings.  I've generated the `2048` bit prime `dhparams` on a variety of machines to see how long they take, there does seem to be a non-linear time component to it:

```
time openssl dhparam -out /tmp/dhparam.pem 2048

Bare-Metal: Intel(R) Core(TM) i7-5960X CPU @ 3.00GHz
0:13        

AWS EC2 t2.micro: Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
1:18

Bare-Metal: Intel(R) Xeon(R) CPU E5-2620 0 @ 2.00GHz
2:26

Rackspace Cloud: Quad-Core AMD Opteron(tm) Processor 2374 HE
6:28

Bare-Metal: Intel(R) Atom(TM) CPU  230   @ 1.60GHz
12:00
```

This literally means that if we add it to the `entrypoint` and start the container on my production Rackspace Cloud instance (which is an HTTP load balancer, btw), it will add *six and a half minutes* to the startup time.  Imagine you update the image and run `docker-compose up --force-recreate` to get the new features, and it takes nearly seven minutes to start :-1:

kamermans commented:
@jwilder: I spent a bunch of time thinking this through and testing things.  The `dhparam` generation time is not predictable.  On my fastest machine, I usually get times like 20 seconds, but sometimes it will take nearly two minutes.  It's really just a matter of chance since `openssl` is looking for large prime numbers arbitrarily.  I think it's conceivable that it could take over 10 minutes on a modern cloud instance (it took 12 min on my old Rackspace instance and 20 min on my Atom 1.6GHz CPU).

As annoying as this generation time is (it's a show-stopper for me), it is important for security reasons, so I went ahead and implemented my 4th idea .  This branch (`kamermans/nginx-proxy:feature_ssl_improvement`) now ships with `/app/dhparam.pem.default`, which I generated on my computer.  On startup, if `/etc/nginx/dhparam/dhparam.pem` is not present, or has the same `md5sum` as `/app/dhparam.pem.default`, a new `2048` bit dhparam is generated in the background with `nice -n +5`, which decreases its CPU priority, since without it, it will peg a CPU core the whole time.  After the generation is complete, nginx is reloaded via `nginx -s reload`.  The process is kicked-off by `docker-entrypoint.sh` (I tried `forego/Procfile`, but I don't see how to run a one-time script).  Since the entrypoint could be called again while the background generation is running, the system uses a lockfile to prevent multiple background generations from running in parallel (and pegging all CPU cores).

I think the whole system should be pretty safe - I've already run it in my environment without issue.

In case someone else wants to give it a spin, I've uploaded it to Docker hub as `kamermans/nginx-proxy:dhparam_bg_init`.  When you first start it, you will see these comments:
```
WARNING: /etc/nginx/dhparam/dhparam.pem was not found. A pregenerated dhparam.pem will be used for now while a new one
is being generated in the background.  Once the new dhparam.pem is in place, nginx will be reloaded.
... normal nginx-proxy init stuff ...
Generating DH parameters, 2048 bit long safe prime, generator 2
This is going to take a long time
dhparam generation complete, reloading nginx
```

Note that there's really no reason we couldn't increase from `2048` bits to `4096` if there is some advantage there, but we could be literally seeing things like hour-long generation times for very slow machines :P

Also, I see that the alpine variant is failing the BATS tests, and I'm not sure why yet, but it's hard to imagine it's related to the dhparam stuff.  I'll take a look at it anyway.

kamermans commented:
I'm not sure why these BATS tests are failing in Travis-CI - it's like the `openssl` command is not running at all or something.  I'll get to the bottom of it.

kamermans commented:
Ok, the Travis-CI environment was not fast enough to generate a `2048` bit dhparam, so I've added an env var `DHPARAM_BITS` which defaults to `2048`, but I set it to `256` for the BATS tests and now it's passing.

kamermans commented:
@jwilder This PR is stable now and I've fixed the merge conflict with a BATS test in `master`.  I think it's ready to be merged if you want.

jwilder commented:
This will need a rebase now that #655 is merged.

kamermans commented:
@jwilder: Rebase complete.  I've put a small hack in place to workaround the Docker version issue, but it should merge fine with #757.

@thomasleveil: please note that I switched the python base image to `python-alpine` so I could get `openssl >= 1.0.2`.  This makes it possible for me to confirm that the `dhparams.pem` file is being used by `nginx` in a unit test.  All tests are passing, so I assume the alpine version is ok.  I'm not a python developer, so perhaps you could take a quick look at `test/test_ssl/test_dhparam*.py` to see if I am doing anything crazy :)

thomasleveil commented on test/test_ssl/test_dhparam.py:
```diff
@@ -0,0 +1,74 @@
+            break
+
+def test_dhparam_is_not_generated_if_present(docker_compose, nginxproxy):
+    wait_for_nginxproxy_to_be_ready()
```
This isn't needed as the `docker_compose` fixture calls it already

kamermans replied:
Ah, so the container is already ready before the tests are run?

thomasleveil replied:
Exactly

thomasleveil replied:
Exactly

thomasleveil commented on test/test_ssl/test_dhparam.yml:
```diff
@@ -0,0 +1,16 @@
+  environment:
+    WEB_PORTS: "85"
+    VIRTUAL_HOST: "web5.nginx-proxy.tld"
+    HTTPS_METHOD: nohttp
```
`HTTPS_METHOD` does not seem to have any effect on the outcome of the test. I would remove it

kamermans replied:
Good point, no need for it here - it was an artifact of me copying another test :)

thomasleveil commented:
@kamermans I've submitted to you kamermans/nginx-proxy/pull/2 with changes that IMHO helps understanding the intents of your tests

kamermans commented:
Very nice, thanks!

kamermans commented:
Please note that at my organization I have run into issues with 2048-bit DH primes.  Java 6 only supported up to 1024-bit DH primes for quite a while, and I have customers that require support for these older versions of Java, so at least in my organization, I have had to move back to 1024-bit.

jwilder commented:
@kamermans I'd lean towards more secure by default so leave the 2048 as default as is.  Did you have any other changes you needed to make before merging?  Looks good to me though.

kamermans commented:
Ok, sounds good to me @jwilder, I have no more changes.

kamermans commented:
@jwilder the merge from `master` caused the tests to fail because `server_names_hash_bucket_size` is now defined in `nginx.tmpl` instead of via `sed` in the `Dockerfile`.  I've fixed that issue and added a blurb to the `README.md` about the DH param compatibility with older clients, as well as a workaround that can be used with these clients (passing `DHPARAM_BITS=1024` to the `nginx-proxy` container).

jwilder commented:
Hmm.  Looks like something on travis or elsewhere is failing now.  Another PR failed for the same reason.

```
1.19s$ sudo apt-get -o Dpkg::Options::="--force-confnew" install -y --force-yes docker-engine=${DOCKER_VERSION}
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Version '1.13.1-0~ubuntu-trusty' for 'docker-engine' was not found
The command "sudo apt-get -o Dpkg::Options::="--force-confnew" install -y --force-yes docker-engine=${DOCKER_VERSION}" failed and exited with 100 during .
```

jwilder commented:
Build is fixed in master.  Thanks @kamermans!
